File: guides to algorithms/Guia08/Python/ArvoreRedBlack/Memoria.py
from io import StringIO
from Pilha import Pilha

class Memoria:

    p = []
    key = []
    left = []
    right = []
    NIL = 0
    
    def __getitem__(self, index):
        return self.s[index]

    def __init__(self, n):
        self.p = [0] * n
        self.key = [0] * n
        self.left = [0] * n
        self.right = [0] * n
        self.s = Pilha(n)
        for k in range(0, n):
            self.s.push(k)

    def allocate_object(self):
        x = self.s.pop()
        if x == 0:
            return 0
        else:
            return x

    def free_object(self, x):
        self.s.push(x)

    def toString(self):
        res = StringIO()
        for k in range((len(self.p)) - 1, -1, -1):
            res.write("{}".format(k))
            res.write("-> [p = ")
            res.write("nil") if self.p[k] == 0 else res.write("{}".format(self.p[k]))
            res.write(", key = ")
            res.write("{}".format(self.key[k]))
            res.write(", left = ")
            res.write("nil") if self.left[k] == 0 else res.write("{}".format(self.left[k]))
            res.write(", right = ")
            res.write("nil") if self.right[k] == 0 else res.write("{}".format(self.right[k]))
            res.write("]\n")
        return res.getvalue()       

File: guides to algorithms/Guia07/Java/ArvorePesquisaBinaria/Memoria.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package arvorepesquisabinaria;

public class Memoria{
    public int[] p = null;
    public int[] key = null;
    public int[] left = null;
    public int[] right = null;
    final int NIL = 0;
    Pilha s = null;
    
    public Memoria(int n){
        this.p = new int[n];
        this.key = new int[n];
        this.left = new int[n];
        this.right = new int[n];
        this.s = new Pilha(n);
        for(int k = 1; k < n; k++){
            this.s.push(k);
        }
        
        for(int k = 0; k < n; k++){
            this.p[k] = this.NIL;
            this.left[k] = this.NIL;
            this.right[k] = this.NIL;
        }
    }
    
    public int allocate_object(){
        int x = this.s.pop();
        if( x == 0 ){
            return 0;
        } else{
            return x;
        }
    }
    
    public void free_object(int x){
        this.s.push(x);
    }
    
    @Override public String toString(){
        StringBuilder res = new StringBuilder();
        res.append("\n");
        for(int k = this.p.length - 1; k >= 0; k--){
            res.append(k);
            res.append("-> [p = ");
            res.append(this.p[k] == 0 ? "nil" : this.p[k] );
            res.append(", key = ");
            res.append(this.key[k]);
            res.append(", left = ");
            res.append(this.left[k] == 0 ? "nil" : this.left[k]);
            res.append(", right = ");
            res.append(this.right[k] == 0 ? "nil" : this.right[k]);
            res.append("]\n");
        }
        return res.toString();
    }
}